plugins/ch341a/fu-ch341a-cfi-device.c

```diff
@@ -0,0 +1,431 @@
+			return NULL;
+		if (i == 0) {
+			g_byte_array_append(blob, buf + 0x4, fu_chunk_get_data_sz(chk) - 0x4);
+		} else {
+			g_byte_array_append(blob, buf + 0x0, fu_chunk_get_data_sz(chk));
+		}
+
+		/* done */
+		fu_progress_step_done(progress);
+	}
+
+	/* success */
+	return g_byte_array_free_to_bytes(g_steal_pointer(&blob));
+}
+
+static gboolean
+fu_ch341a_cfi_device_write_firmware(FuDevice *device,
+				    FuFirmware *firmware,
+				    FuProgress *progress,
+				    FwupdInstallFlags flags,
+				    GError **error)
+{
+	FuCh341aCfiDevice *self = FU_CH341A_CFI_DEVICE(device);
+	FuCh341aDevice *proxy = FU_CH341A_DEVICE(fu_device_get_proxy(FU_DEVICE(self)));
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GBytes) fw_verify = NULL;
+	g_autoptr(GPtrArray) pages = NULL;
+	g_autoptr(FuDeviceLocker) cslocker = NULL;
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* open programmer */
+	locker = fu_device_locker_new(proxy, error);
+	if (locker == NULL)
+		return FALSE;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_ERASE, 33);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 44);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_VERIFY, 35);
+
+	/* get default image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* erase */
+	if (!fu_ch341a_cfi_device_write_enable(self, error)) {
+		g_prefix_error(error, "failed to enable writes: ");
+		return FALSE;
+	}
+	if (!fu_ch341a_cfi_device_chip_erase(self, error)) {
+		g_prefix_error(error, "failed to erase: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* write each block */
+	pages = fu_chunk_array_new_from_bytes(fw, 0x0, 0x0, 0x100);
+	if (!fu_ch341a_cfi_device_write_pages(self,
+					      pages,
+					      fu_progress_get_child(progress),
+					      error)) {
+		g_prefix_error(error, "failed to write pages: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* verify each block */
+	fw_verify = fu_ch341a_cfi_device_read_firmware(self,
+						       g_bytes_get_size(fw),
+						       fu_progress_get_child(progress),
+						       error);
+	if (fw_verify == NULL) {
+		g_prefix_error(error, "failed to verify blocks: ");
+		return FALSE;
+	}
+	if (!fu_common_bytes_compare(fw, fw_verify, error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* success! */
+	return TRUE;
+}
+
+static GBytes *
+fu_ch341a_cfi_device_dump_firmware(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuCh341aCfiDevice *self = FU_CH341A_CFI_DEVICE(device);
+	FuCh341aDevice *proxy = FU_CH341A_DEVICE(fu_device_get_proxy(FU_DEVICE(self)));
+	gsize bufsz = fu_device_get_firmware_size_max(device);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* open programmer */
+	locker = fu_device_locker_new(proxy, error);
+	if (locker == NULL)
+		return NULL;
+
+	/* sanity check */
+	if (bufsz == 0x0) {
```

Comment:
This is out of scope, but I find huge value in that plugin; especially for debugging from very special situations.

However, it is kind of limited because of the "API"; I really like to be able to read and write x bytes at address y.

But, this implies passing those extra parameters, via command line or environment variables and update/add vfunc.

Reply:
For example: I would like to write a part of the blob somewhere in the flash without rewriting the whole flash. but this brings lots of complication. May the thing I would like is just to be able to set the start address.

Reply:
Reading at random offsets is easily possible (and we could certainly change the vfunc if required) but writing is much harder as you typically need to back up any of the affected sector (and possibly block...) then erase, write and verify each sector.

Reply:
Exact; the user has to know what he is doing.